bin/propolis-server/src/lib/stats/virtual_machine.rs

```diff
@@ -0,0 +1,526 @@
+        let vcpu_id = named
+            .iter()
+            .find_map(|named| {
+                if named.name == VCPU_KSTAT_PREFIX {
+                    named.value.as_u32().ok()
+                } else {
+                    None
+                }
+            })
+            .ok_or(Error::NoSuchKstat)?;
+
+        // We track each vCPU microstate starting with `time_`, and map them
+        // into our own definitions of the vCPU states. We need to aggregate all
+        // the occupancy times from the microstates that map to the same public
+        // state.
+        let mut occupancy_by_state = BTreeMap::new();
+        for nv in named.iter() {
+            // Skip kstats that are not known microstate names.
+            let Some(state) = kstat_microstate_to_state_name(nv.name) else {
+                continue;
+            };
+
+            // Get the current summed state occupancy, or insert one with 0.
+            let datum =
+                occupancy_by_state.entry(state.to_string()).or_insert_with(
+                    || Cumulative::with_start_time(*creation_time, 0),
+                );
+            *datum += nv.value.as_u64()?;
+        }
+
+        // Now convert the aggregated occupancy times into samples.
+        for (state, datum) in occupancy_by_state.into_iter() {
+            let metric = VcpuUsage { vcpu_id, state, datum };
+            let sample =
+                Sample::new_with_timestamp(snapshot_time, vm, &metric)?;
+            out.push(sample);
+        }
+    }
+    Ok(out)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::kstat_instance_from_instance_id;
+    use super::produce_vcpu_usage;
+    use super::Data;
+    use super::Kstat;
+    use super::Named;
+    use super::NamedData;
+    use super::Utc;
+    use super::VcpuUsage;
+    use super::VirtualMachine;
+    use super::VCPU_KSTAT_PREFIX;
+    use super::VMM_KSTAT_MODULE_NAME;
+    use super::VM_KSTAT_NAME;
+    use super::VM_NAME_KSTAT;
+    use oximeter::schema::SchemaSet;
+    use oximeter::types::Cumulative;
+    use oximeter::Datum;
+    use oximeter::FieldValue;
+    use uuid::Uuid;
+
+    fn test_virtual_machine() -> VirtualMachine {
+        const INSTANCE_ID: Uuid =
+            uuid::uuid!("96d6ec78-543a-4188-830e-37e2a0eeff16");
+        const PROJECT_ID: Uuid =
+            uuid::uuid!("7b61df02-0794-4b37-93bc-89f03c7289ca");
+        const SILO_ID: Uuid =
+            uuid::uuid!("6a4bd4b6-e9aa-44d1-b616-399d48baa173");
+        VirtualMachine {
+            silo_id: SILO_ID,
+            project_id: PROJECT_ID,
+            instance_id: INSTANCE_ID,
+            n_vcpus: 4,
+            vm_name: INSTANCE_ID.to_string(),
+        }
+    }
+
+    fn test_usage() -> VcpuUsage {
+        VcpuUsage {
+            state: "run".to_string(),
+            vcpu_id: 0,
+            datum: Cumulative::new(100),
+        }
+    }
+
+    #[test]
+    fn test_no_schema_changes() {
```

Comment:
Curiosity about metrics in general, not a problem with this change: is the schema ever allowed to change (e.g. if we were to break things down into a different set of microstates, or report some additional statistics, or something else)? Are metrics versioned? Again, not a problem with this PR, just curious about what the ground rules are here.

Reply:
The short answer is "no", unfortunately.

The longer answer is, "it depends on the changes". I've put a bit of thought into this over the years, but it's getting close to time to handle this robustly and correctly. Here's the current state of affairs, broken down by the kind of change.

It is always fine to report _new_ metrics, such as tracking memory usage or guest VNIC usage. That is a completely new timeseries, and can be done whenever we want. This is done by adding a new type which implements `oximeter::Metric`, and generating samples from it. That could still use the `VirtualMachine` target I've added this PR, just with a new metric.

Breaking things down into a different set of microstates would be allowed, but sort of scary. The onus is currently on the person running to query to inspect when those changes occur and handle them. If you did something like sum across all microstates, then the query is still valid; if you looked at a particular microstate, which was further broken down as part of a change, you would see discontinuities. Nothing would _break_ per se, but you've now got a different set of timeseries.

The most difficult (read: disallowed) change would be adding or removing a field of an existing timeseries. That's what this test is for. Right now, `oximeter` will retrieve your (now changed) samples; notice that its schema is different; report a warning; and helpfully chuck your sample into the bin.

This is really Not Great. I think it's certainly possible to handle the addition of _new_ fields gracefully, by updating the schema; using default or NULL values for old samples; and / or handling the schema change on the query side. The _removal_ of fields or a change in their type is pretty difficult, and I'm not sure how we'd handle it in general. At this still-early stage of development, we can always use the nuclear option of completely expunging the timeseries database or an individual timeseries, during an upgrade. That only works though if you've got one version of the schema deployed.

One thing that could help identify changes is to compute an ID or hash for the schema itself, and then effectively consider two versions of the schema to be two separate timeseries. The query side would then handle running queries against both, or returning an error if that can't be done.

Another way to manage this is to put some API in place in Nexus and / or `oximeter` itself to explicitly manage the update of a schema. `oximeter` could then either expunge old data, or add new (default) values for fields, etc. This is sort of complicated because we store the schema in ClickHouse itself, which has extremely weak consistency guarantees. I've floated the idea of moving schema to Cockroach, but that's a big change that I wouldn't take on without good reason.

Anyway, that's the long and unsatisfying story at this point :)

Reply:
Considering the relative brittleness in the face of changes to the emitted metrics, do you think it would make sense to device them in a module or crate, such that it's clear that any subsequent changes would require the care you mention?

We've started something similar to this with the migration payload data (living in child modules of the various devices, perhaps to be moved to a crate eventually).  The intent there was to make it easier to spot if/when one of those data structures is changed.

Reply:
@pfmooney Could you explain more what that looks like? I'm not entirely sure I follow. Does that mean something like naming the target/metric struct with a version ID, such as `VirtualMachineV1`?

The test here will catch if the schema changes (though not the microstates), since the expected schema file is checked in as part of this PR. It's not really sufficient, but it will at least get more eyes on any change that modifies those schema.

Reply:
> Does that mean something like naming the target/metric struct with a version ID, such as `VirtualMachineV1`?

Not specifically that, no.

> The test here will catch if the schema changes (though not the microstates)

But a change in the reported microstates (or rather what we distill them down into) would be a problem, right?

Having the scheme checked in for the test is good.  I guess we can revisit later if/when we have a good mechanism to constrain the microstates too?

Reply:
A change in the microstates would be confusing, at the very least. The data would continue to be reported and stored in ClickHouse just fine, but it'd be hard to follow, I think. Suppose we decided to break the `"waiting"` microstate into `"waiting_for_this"` and `"waiting_for_that"`. When that change happened, in the database you'd see the end of the timeseries containing the `"waiting"` state (no more samples from it), and the genesis of the two with the other state names (no samples before the change, only after).

I think it's possible to get a consistent data representation in this case, by aggregating the two later states. Since we've explicitly set this up so that their sum is the same as the `"waiting"` state before, that would be OK. But you're certainly correct that doing something that is not a simple, linear mapping between the states before and after the change is much harder to grok.

> I guess we can revisit later if/when we have a good mechanism to constrain the microstates too?

So here's one possible way to check this now:

- Create an array of some number of states, spanning the set of lower-level `kstat` vCPU microstates we expect to produce. Make the number of each state different.
- Iterate through those, and convert to the state we report to `oximeter`.
- Sum the count of each of those
- Assert that is the same as we expect from the input we've created.

This is pretty contrived, but I think it would catch a change in the mapping from `kstat`-layer states into `oximeter`-layer. That seems worth doing, no?

Reply:
Ok, I've added a test, which I'll include in the next commit. It looks tautological, but I think it's useful. It catches the following:

- Removing one of the two kstat-states that map to the oximeter `"idle"` state
- Mapping that state to a new oximeter state
- Mapping both of those states to a single oximeter state, but with a different name.
- Mapping a new kstat state into an existing oximeter state

The one thing it does not catch is if the `kstat` layer grows a new state, or removes one. This just checks the mapping between the two. We could certainly add such a test, provided we can spawn VMs so that the `kstat`s show up in userland.

Reply:
This test was added in 7c4cbe7